impl/src/as/mod.rs

```diff
@@ -0,0 +1,371 @@
+    (trait_ident, method_ident, conv_type, mutability): ExpansionCtx<'_>,
+) -> syn::Result<TokenStream> {
+    let trait_path = quote! { ::derive_more::#trait_ident };
+
+    let field_args = extract_field_args(input, trait_ident, method_ident)?;
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+    let input_type = &input.ident;
+
+    Ok(
+        field_args
+            .into_iter()
+            .map(
+                |FieldArgs {forward, field, ident}| {
+                    let member = quote! { self.#ident };
+                    let field_type = &field.ty;
+                    let (
+                        body,
+                        impl_generics,
+                        where_clause,
+                        trait_path,
+                        return_type
+                    ) = if forward {
+                        let trait_path = quote! { #trait_path<#conv_type> };
+
+                        let with_constraint = add_extra_where_clauses(
+                            &input.generics,
+                            quote! {
+                                where #field_type: #trait_path
+                            },
+                        );
+                        let new_generics = add_extra_generic_type_param(
+                            &with_constraint,
+                            quote! {
+                                #conv_type: ?::core::marker::Sized
+                            },
+                        );
+                        let (impl_generics, _, where_clause) =
+                            new_generics.split_for_impl();
+
+                        let casted_trait = quote! { <#field_type as #trait_path> };
+
+                        (
+                            quote! { #casted_trait::#method_ident(& #mutability #member) },
+                            quote! { #impl_generics },
+                            quote! { #where_clause },
+                            quote! { #trait_path },
+                            quote! { & #mutability #conv_type },
+                        )
+                    } else {
+                        (
+                            quote! { & #mutability #member },
+                            quote! { #impl_generics },
+                            quote! { #where_clause },
+                            quote! { #trait_path<#field_type> },
+                            quote! { & #mutability #field_type },
+                        )
+                    };
+
+                    quote! {
+                        #[automatically_derived]
+                        impl #impl_generics #trait_path for #input_type #ty_generics #where_clause {
+                            #[inline]
+                            fn #method_ident(& #mutability self) -> #return_type {
+                                #body
+                            }
+                        }
+                    }
+                })
+            .collect()
```

Comment:
@MegaBluejay I don't really like the solution here. It somewhat inverts how and where we handle the abstractions in the code. The type representing an expansion is actually expanded in the main `expand` function, while parsing/forming those expansions is made in separate functions below. It's more natural to see the opposite: we parse and form the expansion values in the main `expand` function, while the details of parsing and how exactly expansion is rendered - are abstracted behind the types and theirs APIs.

This made me to rewrite the whole thing. Please, examine the new code carefully, and consider/use this decomposition patter for future contributions.